optional.py

```diff
@@ -31,6 +33,21 @@ def get(self):
 
         return self._thing
 
+    def map(self, func):
```

Comment:
This brings up an interesting point about purity vs. practicality. It's possible in Python to combine these methods. To always do 'the right thing' since we are not constrained by static types. This is what JavaScript does with it's Promise implementation. I'm not for it, but it's worth a conversation, especially as a separate named method.

Reply:
yeah I was wondering about that. I don't like code that assumes to know how someone wanted to use it.  On the other hand I don't like my type checking (it feels like a violation of ducktyping) but I couldn't think of a better way.